File: dynamic programming/fibonacci_dynamic.js
// 0, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233

let calc = 0;
function fibonacciDynamic() {
    const cache = {};
    return function fib(n) {
        calc++;
        if (n in cache) {
            return cache[n];
        } else {
            if (n < 2) {
                return n;
            } else {
                cache[n] = fib(n-1) + fib(n-2);
                return cache[n];
            }
        }
    }
}

const fib = fibonacciDynamic();
console.log("DP", fib(69)); // O(n)
console.log("We did " + calc + " calculations!");

File: dynamic programming/fibonacci_dynamic.lua
local calc = 0;

local function fibonacciDynamic()
    local cache = {}
    local function fib(n)
        calc = calc + 1;
        if cache[n] then
            return cache[n]
        else
            if n < 2 then
                return n
            else
                cache[n] = fib(n - 1) + fib(n - 2)
                return cache[n]
            end
        end
    end
    return fib;
end

local fib = fibonacciDynamic();
print("DP " .. fib(69) .. "!") -- O(n)
print("We did " .. calc .. " calculations!")

